### DISCUSSION TITLE
Smaller improvements

### DISCUSSION THREAD

**Discussion by klytje** [AUSAXS/AUSAXS#98] [Ideas]
Discussion for smaller improvements

Current list:
* [x] Symmetries on _groups_ of bodies. This can perhaps be accomplished by allowing pointer-to-symmetry members of each body.
* [x] Support tetrahedral symmetries.
* [ ] Support end-of-line comments in the rigidbody script. 
* [ ] Clean up everything related to global settings. This includes getters, setters, and how they are being stored. 
* [ ] Add `reset_settings` method for resetting all settings to their default values. This is useful for both testing and through the python API, where it would allow settings to be set one-by-one instead of all-at-once.  
* [ ] Due to the binning approximation, we only support a limited number of form factors, though the calculations do not really care which are used. We should therefore consider allowing a few of them to be user-replacable, to better support exchanging unused form factors for e.g. more accurate heavy element calculations. 
* [ ] `AtomFF` probably should not have a dedicated `weight` field, but rather dynamically fetch it when requested. 
* [x] Clean up `CompactCoordinatesTemplate` constructors. 
* [ ] `debye_transform(vector q)` is only defined for the base `DistanceHistogram` calculator.
* [ ] `settings::flags` should perhaps be renamed to something a la `internal_state` since they have been expanded from pure bools and shouldn't be touched. 
* [ ] SmartFitter: avoid inverting dataset for every iteration
* [ ] EM: estimated computation time
* [ ] RigidBody: estimated computation time
* [ ] FormFactor: implement actual dummy water form factor. Currently OH is being used. 
* [ ] PartialHistogramManager: Shift BodyTracker responsibility to the Molecule itself, and make the constructor attach a listener object to the Molecule for updates instead. 
* [ ] Math: Add array-based Matrix3<T> to allow loop unrolling of rotations. 
* [ ] HistogramManagers should be consistent in their handling of cross terms. The base HistogramManager and HistogramManagerMT scale the cross terms by a factor 2, while the remaining ones do not. 
* [ ] LinearLeastSquares: Consider using references to data instead of enforcing copying 
* [ ] `WeightedEntry::increment` performs unnecessary calculations (incrementing both count & value). We can probably branch before evaluating the weighted bin centers depending on if the value is empty or not. 
* [x] Benchmark histogram calculation vs intensity calculation. I suspect the latter is significantly slower. 
* [ ] Support dynamic q-axes for all excluded volume models. Currently, only the Simple model supports this. The solution is to make `debye_transform(q)` do some axis setup, and then delegate to `debye_transform()`, which would make the workflow agnostic to which histogram model is currently selected. The main difficulty here is that `sinqd` tables for custom q-axes are indexed differently, leading to invalid indexing. 
* [ ] Lost opportunity for further optimization of `ReferenceSymmetries`?
* [ ] Consider removing the separate setup phase when running rigid-body refinements. This was only implemented to avoid errors when writing scripts programmatically; an approach increasingly disfavored due to the heavy focus on text scripting. 

GUI stuff:
* [ ] EM: Warn user if existing files will be overwritten
* [ ] SAXS: Warn user if existing files will be overwritten
* [ ] Make overall GUI program containing both EM and SAXS as panes?
* [ ] GUIs: Inform user of plotting issues instead of throwing an exception.

**Comment by Marcslh14** [comment]
Hej Kristian, I hope all is going well. Do you think it would be difficult to calculate the theoretical scattering of a pdb structure? No experimental data, just the structure itself with an arbitrary (constant) concentration, then one can try linear combinations. This we do it with Jan's software but with yours it could be more user friendly and probably shouldn't be too difficult (?).
Best 
Marcos

**Reply by klytje** [reply]
Hi Marcos! 

Yes, you can actually do this already. Just run the `saxs_fitter` with only the pdb file & it will spit out the theoretical scattering. The only issue is the hydration shell: our model cannot use "real" densities, and the initial hydration contribution is therefore arbitrarily scaled. Usually this is not a problem, as an overall density scaling factor is fitted to the data anyways. But without a dataset to compare against, its contribution becomes unphysical. 

Three solutions. 
1. If you're only interested in low-q agreement, probably you can just neglect the hydration shell entirely. 
2. You could extract the calculated "aw" and "ww" partial intensity profiles (currently only available through the profiles.png.plot output file) and add the missing scaling factor yourself as an additional free parameter (i.e. weighting the "aw" profile by `c`, and the "ww" profile by `c^2`). Not a great solution, but it should get the job done. 
3. If your structures are not too different, probably the hydration shell should contribute more or less equally for all of them. If so, you can fit a few of them individually to the data, take the fitted `c` values, and do the scaling in (2) with these. I will add an option for doing this through the command-line soon, now that I realize it is a useful feature. 

As I write this, I think it would be good to add better support for this use-case; I will look at improving this in the not-too-distant future. Let me know if you need more help and/or information.
